torchmultimodal/modules/encoders/swin_transformer_3d_encoder.py

```diff
@@ -0,0 +1,538 @@
+        self.relative_position_bias_table = nn.Parameter(
+            torch.zeros(
+                (2 * window_size[0] - 1)
+                * (2 * window_size[1] - 1)
+                * (2 * window_size[2] - 1),
+                num_heads,
+            )
+        )  # 2*Wd-1 * 2*Wh-1 * 2*Ww-1, nH
+
+        # get pair-wise relative position index for each token inside the window
+        coords_dhw = [torch.arange(self.window_size[i]) for i in range(3)]
+        coords = torch.stack(
+            torch.meshgrid(coords_dhw[0], coords_dhw[1], coords_dhw[2], indexing="ij")
+        )  # 3, Wd, Wh, Ww
+        coords_flatten = torch.flatten(coords, 1)  # 3, Wd*Wh*Ww
+        relative_coords = (
+            coords_flatten[:, :, None] - coords_flatten[:, None, :]
+        )  # 3, Wd*Wh*Ww, Wd*Wh*Ww
+        relative_coords = relative_coords.permute(
+            1, 2, 0
+        ).contiguous()  # Wd*Wh*Ww, Wd*Wh*Ww, 3
+        relative_coords[:, :, 0] += self.window_size[0] - 1  # shift to start from 0
+        relative_coords[:, :, 1] += self.window_size[1] - 1
+        relative_coords[:, :, 2] += self.window_size[2] - 1
+
+        relative_coords[:, :, 0] *= (2 * self.window_size[1] - 1) * (
+            2 * self.window_size[2] - 1
+        )
+        relative_coords[:, :, 1] *= 2 * self.window_size[2] - 1
+        relative_position_index = relative_coords.sum(-1)  # Wd*Wh*Ww, Wd*Wh*Ww
+        self.register_buffer("relative_position_index", relative_position_index)
+
+        nn.init.trunc_normal_(self.relative_position_bias_table, std=0.02)
+
+    def forward(self, x: Tensor):
+        _, D, H, W, _ = x.shape
+        size_dhw = (D, H, W)
+        window_size, shift_size = self.window_size.copy(), self.shift_size.copy()
+        # Handle case where window_size is larger than the input tensor
+        for i in range(3):
+            if size_dhw[i] <= window_size[i]:
+                # In this case, window_size will adapt to the input size, and no need to shift
+                window_size[i] = size_dhw[i]
+                shift_size[i] = 0
+
+        N = window_size[0] * window_size[1] * window_size[2]
+        relative_position_bias = self.relative_position_bias_table[self.relative_position_index[:N, :N].reshape(-1)]  # type: ignore[index]
+        relative_position_bias = relative_position_bias.view(N, N, -1)
+        relative_position_bias = (
+            relative_position_bias.permute(2, 0, 1).contiguous().unsqueeze(0)
+        )
+
+        return shifted_window_attention_3d(
+            x,
+            self.qkv.weight,
+            self.proj.weight,
+            relative_position_bias,
+            window_size,
+            self.num_heads,
+            shift_size=shift_size,
+            attention_dropout=self.attention_dropout,
+            dropout=self.dropout,
+            qkv_bias=self.qkv.bias,
+            proj_bias=self.proj.bias,
+        )
+
+
+class SwinTransformerBlock(nn.Module):
```

Comment:
Given you made the necessary changes on TorchVision to support other attention types (https://github.com/pytorch/vision/pull/6088), could you use those components instead of copy-pasting?

Reply:
You are right, will upstream the components from torchvision.